conngen/conngen.cpp

```diff
@@ -132,86 +50,278 @@ nest::cg_connect( nest::ConnectionGeneratorDatum& cg,
-nest::cg_select_implementation( const StringDatum& library,
-  const StringDatum& tag )
+cg_set_masks( ConnectionGeneratorDatum& cg,
+  RangeSet& sources,
+  RangeSet& targets )
 {
-  ConnectionGenerator::selectCGImplementation( tag, library );
+  long np = kernel().mpi_manager.get_num_processes();
+  std::vector< ConnectionGenerator::Mask > masks(
+    np, ConnectionGenerator::Mask( 1, np ) );
+
+  cg_create_masks( &masks, sources, targets );
+  cg->setMask( masks, kernel().mpi_manager.get_rank() );
 }
 
+/**
+ * Create the masks for sources and targets based on the contiguous
+ * ranges given in sources and targets. We need to do some index
+ * translation here, as the CG expects indices from 0..n for both
+ * source and target populations, while the RangeSets sources and
+ * targets contain NEST global indices (gids).
+ *
+ * The masks for the sources must contain all nodes (local+remote).
+ * The skip of the mask was set to 1 in cg_set_masks(). The same
+ * source mask is stored n_proc times on each process.
+ *
+ * The masks for the targets must only contain local nodes. This is
+ * achieved by first setting skip to num_processes upon creation of
+ * the mask in cg_set_masks(), and second by the fact that for each
+ * contiguous range of nodes in a mask, each of them contains the
+ * index-translated id of the first local neuron as the first
+ * entry. If this renders the range empty (i.e. because the first
+ * local id is beyond the last element of the range), the range is
+ * not added to the mask.
+ *
+ * \param masks The std::vector of Masks to populate
+ * \param sources The source ranges to create the source masks from
+ * \param targets The target ranges to create the target masks from
+ *
+ * \note Each process computes the full set of source and target
+ * masks, i.e. one mask per rank will be created on each rank.
+ *
+ * \note Setting the masks for all processes on each process might
+ * become a memory bottleneck when going to very large numbers of
+ * processes. Especially so for the source masks, which are all the
+ * same. This could be solved by making the ConnectionGenerator
+ * interface MPI aware and communicating the masks during connection
+ * setup.
+*/
 void
-nest::cg_set_masks( nest::ConnectionGeneratorDatum& cg,
-  IntVectorDatum& sources,
-  IntVectorDatum& targets )
+cg_create_masks( std::vector< ConnectionGenerator::Mask >* masks,
+  RangeSet& sources,
+  RangeSet& targets )
 {
-  RangeSet source_ranges;
-  cg_get_ranges( source_ranges, ( *sources ) );
+  // The index of the left border of the currently looked at range
+  // (counting from 0). This is used for index translation.
+  size_t cg_idx_left = 0;
 
-  RangeSet target_ranges;
-  cg_get_ranges( target_ranges, ( *targets ) );
+  // For sources, we only need to translate from NEST to CG indices.
+  for ( RangeSet::iterator source = sources.begin(); source != sources.end();
+        ++source )
+  {
+    size_t num_elements = source->last - source->first;
```

Comment:
I dislike both of your suggestions for adding a suffix to the variable name. Instead I've added `+1`/`-1` to the assignment and uses of the variable in 292f690 to make things still work and keep the simple name.

Regarding the other points  by @heplesser I'm sorry, but I think that this is not the place to discuss the design of libneurosim or the Connection Generator Interface. Please feel free to open corresponding issues or pull requests here: https://github.com/INCF/libneurosim/.